parser/src/format/avro.rs

```diff
@@ -0,0 +1,409 @@
+                })?;
+            Ok(Value::String(ts.date().to_string()))
+        }
+        TimeMillis(t) if allow_string_repr => {
+            time_from_midnight("time-millis", column_name, t as i64, MILLIS_PER_SEC)
+        }
+        TimeMicros(t) if allow_string_repr => {
+            time_from_midnight("time-micros", column_name, t, MICROS_PER_SEC)
+        }
+        TimestampMillis(t) if allow_string_repr => {
+            timestamp_from_unix_epoch("timestamp-millis", column_name, t, MILLIS_PER_SEC)
+        }
+        TimestampMicros(t) if allow_string_repr => {
+            timestamp_from_unix_epoch("timestamp-micros", column_name, t, MICROS_PER_SEC)
+        }
+        // If !allow_string_repr, then all the date values will be converted directly to json
+        // numbers. This allows users to handle any conversions themselves, by disallowing string
+        // types in their json schema.
+        Date(i) | TimeMillis(i) => Ok(Value::Number(i.into())),
+        TimeMicros(i) | TimestampMicros(i) | TimestampMillis(i) => Ok(Value::Number(i.into())),
+
+        Duration(avro_dur) => {
+            // avro durations are really weird. We always convert them to json objects, since
+            // there's no obvious alternative representation because they encode a number of
+            // _months_. I seriously doubt that anyone uses this goofy crap anyway.
+            Ok(serde_json::json!({
+                "months": u32::from(avro_dur.months()),
+                "days": u32::from(avro_dur.days()),
+                "millis": u32::from(avro_dur.millis()),
+            }))
+        }
+        Uuid(uuid) => Ok(Value::String(uuid.to_string())),
+        // The decimal type in avro_rs does't allow access to the underlying value (which seems
+        // like a bug). So we just use the built-in converstion, which results in a json array of
+        // the underlying byte values. It probably doesn't matter much, since probably not many
+        // people are using these anyway.
+        other @ Decimal(_) => Ok(std::convert::TryFrom::try_from(other)
+            .expect("converting decimal value to json is infallible")),
+    }
+}
+
+const MILLIS_PER_SEC: i64 = 1_000;
+const MICROS_PER_SEC: i64 = 1_000_000;
+const NANOS_PER_SEC: i64 = 1_000_000_000;
+
+fn time_from_midnight(
+    avro_type: &'static str,
+    column: &str,
+    duration: i64,
+    units_per_sec: i64,
+) -> Result<Value, AvroError> {
+    // times don't allow negative values
+    let time = Some(to_secs_and_nanos(duration, units_per_sec))
+        .filter(|&(secs, _)| secs >= 0)
+        .and_then(|(secs, nanos)| NaiveTime::from_num_seconds_from_midnight_opt(secs as u32, nanos))
+        .ok_or_else(|| AvroError::DateTimeOverflow(duration, avro_type, column.to_string()))?;
+
+    Ok(Value::String(time.to_string()))
+}
+
+fn to_secs_and_nanos(duration: i64, units_per_sec: i64) -> (i64, u32) {
```

Comment:
Not exactly sure on the function requirement. Is it a correct function call of `to_secs_and_nanos(1001, 1000)`? And what is the expect output for that?

Reply:
I'm really glad you asked! I went to add a comment, and spotted a heinous bug where negative timestamps weren't handled correctly. The `chrono` crate expects the sub-second nanos value to always be positive, so if a timestamp is negative, you need to subtract an additional second, and then add back in a the proportional number of sub-second nanos. I fixed that and added a test. Also added a comment explaining what's going on here.

Reply:
Looks great! :+1: